NDB.Covid19/NDB.Covid19/WebServices/AnonymousTokenService.cs

```diff
@@ -0,0 +1,163 @@
+using Org.BouncyCastle.Crypto.EC;
+using Org.BouncyCastle.Crypto.Parameters;
+using Org.BouncyCastle.Math;
+using Org.BouncyCastle.Math.EC;
+using Org.BouncyCastle.Utilities.Encoders;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NDB.Covid19.WebServices
+{
+    public class AnonymousTokenService
+    {
+        private Initiator _initiator = new Initiator();
+        private X9ECParameters _ecParameters;
+
+        public AnonymousTokenService(X9ECParameters ecParameters)
+        {
+            _ecParameters = ecParameters;
+        }
+
+        public async Task<string> GetAnonymousTokenAsync()
+        {
+            var tokenState = GenerateTokenRequest();
+            var tokenResponse = await RequestTokenGeneration(tokenState);
+            var publicKey = await GetPublicKeyAsync(tokenResponse);
+            var token = RandomizeToken(tokenState, tokenResponse, publicKey);
+            return EncodeToken(tokenState, token, tokenResponse.Kid);
+        }
+
+        public AnonymousTokenState GenerateTokenRequest()
+        {
+            var init = _initiator.Initiate(_ecParameters.Curve);
+            return new AnonymousTokenState
+            {
+                t = init.t,
+                r = init.r,
+                P = init.P
+            };
+        }
+
+        public ECPoint RandomizeToken(AnonymousTokenState state, GenerateTokenResponseModel tokenResponse, ECPublicKeyParameters K)
+        {
+            var Q = _ecParameters.Curve.DecodePoint(Hex.Decode(tokenResponse.QAsHex));
+            var c = new BigInteger(Hex.Decode(tokenResponse.ProofCAsHex));
+            var z = new BigInteger(Hex.Decode(tokenResponse.ProofZAsHex));
+            return _initiator.RandomiseToken(_ecParameters, K, state.P, Q, c, z, state.r);
+        }
+
+        string EncodeToken(AnonymousTokenState tokenState, ECPoint W, string keyId)
+        {
+            var t = tokenState.t;
+            return Convert.ToBase64String(W.GetEncoded()) + "." + Convert.ToBase64String(t) + "." + keyId;
+        }
+
+        private async Task<GenerateTokenResponseModel> RequestTokenGeneration(AnonymousTokenState state)
+        {
+            var request = new HttpRequestMessage(HttpMethod.Post, OAuthConf.OAUTH2_ANONTOKEN_URL);
+            request.Headers.Add("Authorization", $"Bearer {AuthenticationState.PersonalData?.Access_token}");
+            var tokenRequest = new GenerateTokenRequestModel
+            {
+                PAsHex = Hex.ToHexString(state.P.GetEncoded())
+            };
+            request.Content = new StringContent(JsonConvert.SerializeObject(tokenRequest), Encoding.UTF8, "application/json");
+            var response = await new HttpClient().SendAsync(request);
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new Exception("Call to " + OAuthConf.OAUTH2_ANONTOKEN_URL + " failed: " + response);
```

Comment:
Suggest we throw explicit custom exception and not generic ones.

Reply:
I agree. Is there an existing set of exceptions that should be used?

Reply:
Look under Models/UserDefinedExceptions for the shared project if one of those are suitable, otherwise define your own there.